File: swig/test/test_constructors.py
#
#   Test file for HfstTransducer constructors, destructor, operator=
#   and member functions set_name, get_name and get_type.
#

import libhfst

types = [libhfst.SFST_TYPE, libhfst.TROPICAL_OPENFST_TYPE, libhfst.FOMA_TYPE]

for type in types:

    if not libhfst.HfstTransducer.is_implementation_type_available(type):
        continue

    # The empty transducer
    print("The empty transducer")
    empty = libhfst.HfstTransducer(type)

    # The epsilon transducer
    print("The epsilon transducer")
    epsilon = libhfst.HfstTransducer("@_EPSILON_SYMBOL_@", type)
    
    # One-transition transducer
    print("One-transition transducer")
    foo = libhfst.HfstTransducer("foo", type)
    foobar = libhfst.HfstTransducer("foo", "bar", type)
    
    # The copy constructor
    print("The copy constructor")
    foobar_copy = libhfst.HfstTransducer(foobar)
    assert(foobar.compare(foobar_copy))
    
    # Conversion from HfstBasicTransducer
    print("Conversion from HfstBasicTransducer")
    basic = libhfst.HfstBasicTransducer()
    basic.add_state(1)
    basic.add_transition(0, libhfst.HfstBasicTransition(1, "foo", "bar", 0))
    basic.set_final_weight(1, 0)
    foobar_basic = libhfst.HfstTransducer(basic, type)
    assert(foobar.compare(foobar_basic))
    
    # By tokenizing
    print("Construction by tokenization")
    tok = libhfst.HfstTokenizer()
    tok.add_skip_symbol("baz")
    tok.add_multichar_symbol("foo")
    tok.add_multichar_symbol("bar")
    foo_tok = libhfst.HfstTransducer("bazfoobaz", tok, type)
    foobar_tok = libhfst.HfstTransducer("bazfoo", "barbaz", tok, type)
    assert(foo.compare(foo_tok))
    assert(foobar.compare(foobar_tok))
        
    # Function assign
    print("Function assign()")
    foobar2 = libhfst.HfstTransducer("baz", type)
    foobar.set_name("foobar")
    assert(foobar.get_name() == "foobar")
    foobar2.assign(foobar)
    assert(foobar2.get_name() == "foobar")
    assert(foobar.compare(foobar2))
    empty_ol = libhfst.HfstTransducer(libhfst.HFST_OL_TYPE)
    empty_olw = libhfst.HfstTransducer(libhfst.HFST_OLW_TYPE)
    # reserving props in copy constructor (bug: #3405831)
    t = libhfst.HfstTransducer("a", libhfst.TROPICAL_OPENFST_TYPE)
    t.convert(libhfst.HFST_OLW_TYPE)
    t.set_name("foo")
    s = libhfst.HfstTransducer(t)
    assert(s.get_name() == t.get_name())
    try:
        empty_ol = foobar2.convert(libhfst.HFST_OL_TYPE)
        empty_olw = foobar2.convert(libhfst.HFST_OLW_TYPE)
        assert(empty_ol.get_name() == "foobar")
        assert(empty_olw.get_name() == "foobar")
    except: # libhfst.FunctionNotImplementedException:
        assert(False)
    

File: test/libhfst/test_constructors.cc
/*
   Test file for HfstTransducer constructors, destructor, operator=
   and member functions set_name, get_name and get_type.
*/

#include "HfstTransducer.h"
#include "auxiliary_functions.cc"

using namespace hfst;
using hfst::implementations::HfstBasicTransducer;
using hfst::implementations::HfstBasicTransition;


int main(int argc, char **argv) 
{

  const unsigned int TYPES_SIZE=3;
  const ImplementationType types [] = {SFST_TYPE, 
                       TROPICAL_OPENFST_TYPE, 
                       /*LOG_OPENFST_TYPE,*/ 
                       FOMA_TYPE};

  for (unsigned int i=0; i<TYPES_SIZE; i++)
    {
      if (not HfstTransducer::is_implementation_type_available(types[i]))
	continue;

      /* The empty transducer. */
      verbose_print("The empty transducer", types[i]);
      HfstTransducer empty(types[i]);

      /* The epsilon transducer. */
      verbose_print("The epsilon transducer", types[i]);
      HfstTransducer epsilon("@_EPSILON_SYMBOL_@", types[i]);

      /* One-transition transducer. */
      verbose_print("One-transition transducer", types[i]);
      HfstTransducer foo("foo", types[i]);
      HfstTransducer foobar("foo", "bar", types[i]);

      /* The copy constructor. */
      verbose_print("The copy constructor", types[i]);
      HfstTransducer foobar_copy(foobar);
      assert(foobar.compare(foobar_copy));

      /* Conversion from HfstBasicTransducer. */
      verbose_print("Conversion from HfstBasicTransducer", types[i]);
      HfstBasicTransducer basic;
      basic.add_state(1);
      basic.add_transition(0, HfstBasicTransition(1, "foo", "bar", 0));
      basic.set_final_weight(1, 0);
      HfstTransducer foobar_basic(basic, types[i]);
      assert(foobar.compare(foobar_basic));

      /* By tokenizing. */
      verbose_print("Construction by tokenization", types[i]);
      HfstTokenizer tok;
      tok.add_skip_symbol("baz");
      tok.add_multichar_symbol("foo");
      tok.add_multichar_symbol("bar");
      HfstTransducer foo_tok("bazfoobaz", tok, types[i]);
      HfstTransducer foobar_tok("bazfoo", "barbaz", tok, types[i]);
      assert(foo.compare(foo_tok));
      assert(foobar.compare(foobar_tok));

      /* From AT&T format. */
      verbose_print("Construction from AT&T format", types[i]);
      FILE * file = fopen((std::string(getenv("srcdir")) + 
               std::string("/foobar.att")).c_str(), "rb");
      assert(file != NULL);
      unsigned int linecount = 0;
      HfstTransducer foobar_att(file, types[i], "@0@", linecount);
      (void)linecount;
      fclose(file);
      foobar_att.minimize();

      assert(foobar.compare(foobar_att));

      /* From HfstInputStream. 
     Tests also functions get_type, set_name and get_name. */
      verbose_print("Construction from HfstInputStream", types[i]);
      HfstOutputStream out("testfile.hfst", foobar.get_type());
      foobar.set_name("foobar");
      out << foobar;
      out.close();
      HfstInputStream in("testfile.hfst");
      HfstTransducer foobar_stream(in);
      in.close();
      remove("testfile.hfst");
      assert(foobar.compare(foobar_stream));
      assert(foobar_stream.get_name().compare("foobar") == 0);
      assert(foobar_stream.get_type() == types[i]);

      /* Destructor. */
      verbose_print("Destructor", types[i]);
      HfstTransducer * nu = new HfstTransducer("new", types[i]);
      delete nu;

      /* Operator=. */
      verbose_print("Operator=", types[i]);
      HfstTransducer foobar2("baz", types[i]);
      assert(foobar.get_name().compare("foobar") == 0);
      foobar2 = foobar;
      assert(foobar2.get_name().compare("foobar") == 0);
      assert(foobar.compare(foobar2));
      HfstTransducer empty_ol(HFST_OL_TYPE);
      HfstTransducer empty_olw(HFST_OLW_TYPE);
      /* reserving props in copy constructor (bug: #3405831) */
      HfstTransducer t("a",TROPICAL_OPENFST_TYPE);
      t.convert(HFST_OLW_TYPE);
      t.set_name("foo");
      HfstTransducer s(t);
      assert(s.get_name() == t.get_name());
      try {
        empty_ol = foobar2.convert(HFST_OL_TYPE);
        empty_olw = foobar2.convert(HFST_OLW_TYPE);
	assert(empty_ol.get_name().compare("foobar") == 0);
	assert(empty_olw.get_name().compare("foobar") == 0);
      }
      catch (const FunctionNotImplementedException e)
	{
	  assert(false);
	}

    }

}
